src/gui/widgets/LinkedModelGroupLayout.cpp

```diff
@@ -0,0 +1,303 @@
+**   * Neither the name of The Qt Company Ltd nor the names of its
+**     contributors may be used to endorse or promote products derived
+**     from this software without specific prior written permission.
+**
+**
+** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+** "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+** LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+** A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+** OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+** LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+** DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+** THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
+**
+** $QT_END_LICENSE$
+**
+****************************************************************************/
+
+
+#include "LinkedModelGroupLayout.h"
+
+#include <QWidget>
+#include <QLayoutItem>
+#include <QLineEdit>
+#include <QRect>
+#include <QString>
+
+// TODO: rename: ModelGroupLayout
+
+constexpr const int LinkedModelGroupLayout::m_minWidth;
+
+LinkedModelGroupLayout::LinkedModelGroupLayout(QWidget *parent, int margin, int hSpacing, int vSpacing)
+	: QLayout(parent), m_hSpace(hSpacing), m_vSpace(vSpacing),
+	  m_searchBar(new QLineEdit(parent))
+{
+	setContentsMargins(margin, margin, margin, margin);
+	m_searchBar->setPlaceholderText("filter");
+	connect(m_searchBar, SIGNAL(textChanged(const QString&)),
+		this, SLOT(onTextChanged(const QString& )));
+	addWidget(m_searchBar);
+	m_searchBar->setHidden(true); // nothing to filter yet
+}
+
+LinkedModelGroupLayout::~LinkedModelGroupLayout()
+{
+	QLayoutItem *item;
+	while ((item = takeAt(0))) { delete item; }
+}
+
+void LinkedModelGroupLayout::onTextChanged(const QString&)
+{
+	invalidate();
+	update();
+}
+
+void LinkedModelGroupLayout::addItem(QLayoutItem *item)
+{
+	QWidget* widget = item->widget();
+	const QString str = widget ? widget->objectName() : QString("unnamed");
+	m_itemMap.insert(str, item);
+	invalidate();
+}
+
+int LinkedModelGroupLayout::horizontalSpacing() const
+{
+	if (m_hSpace >= 0) { return m_hSpace; }
+	else
+	{
+		return smartSpacing(QStyle::PM_LayoutHorizontalSpacing);
+	}
+}
+
+int LinkedModelGroupLayout::verticalSpacing() const
+{
+	if (m_vSpace >= 0) { return m_vSpace; }
+	else
+	{
+		return smartSpacing(QStyle::PM_LayoutVerticalSpacing);
+	}
+}
+
+int LinkedModelGroupLayout::count() const
+{
+	return m_itemMap.size() - 1;
+}
+
+QMap<QString, QLayoutItem*>::const_iterator
+LinkedModelGroupLayout::pairAt(int index) const
+{
+	if (index < 0)
+		return m_itemMap.cend();
+	QMap<QString, QLayoutItem*>::const_iterator itr = m_itemMap.cbegin();
+	++itr; // skip search bar
```

Comment:
Without digging through what this means and what's connected to this, what if search bar isn't the only thing we want to skip later on?

Reply:
If we add more generic widgets like search bar, we'll add more `++itr`, e.g. `++++itr`

Reply:
There should be a better way to handle this then.

Reply:
Like checking if the first widget is really a search bar? Or what do you mean?

Reply:
If we need to skip something, maybe the object should tell us we don't need to check it?

Reply:
Reason being is because say we add a lot more of these, then it turns into `++++++++++++itr;` when it should really be something like `while (itr != m_itemMap.end() && itr->skipMe()) { ++itr; }`

Reply:
I could give the search bar a special object name (`QObject::setObjectName`), like `__searchBar__`, and then rewrite the loop as

```
auto skip = [&](QLayoutItem* item) -> bool
{
	return item->widget()->objectName() == "__searchBar__";
};
for (; itr != m_itemMap.cend() && (index > 0 || skip(itr.value())); ++itr)
{
	if(!skip(itr.value())) { index--; }
}
```

Looks a bit hard to read though... Is that what you mean?

Reply:
Yes, this is what I mean essentially. Instead of this though it should be the widget telling us if it needs to be skipped for some reason. So if I'm the model group and I'm going through all the models, I'd ask you "Do I need to skip you?" and you should tell me yes. Alternatively when a model is added it should tell the group that it needs to be skipped, that way we know ahead of time which to skip.

Also, do we use the object name for anything? If not, then it could work, but for future-proofing we should prepend something like `__skip_model_group__` or whatever and check for that.

Reply:
> Instead of this though it should be the widget telling us if it needs to be skipped for some reason

That's why the code has `item->widget()`. This is exactly the widget.

> we should prepend something like __skip_model_group__

I can use `__skip_model_group__` as object name